src/Mvc/Mvc.Core/src/Routing/ControllerRequestDelegateFactory.cs

```diff
@@ -0,0 +1,101 @@
+
+        public ControllerRequestDelegateFactory(
+            ControllerActionInvokerCache controllerActionInvokerCache,
+            IOptions<MvcOptions> optionsAccessor,
+            ILoggerFactory loggerFactory,
+            DiagnosticListener diagnosticListener,
+            IActionResultTypeMapper mapper)
+            : this(controllerActionInvokerCache, optionsAccessor, loggerFactory, diagnosticListener, mapper, null)
+        {
+        }
+
+        public ControllerRequestDelegateFactory(
+            ControllerActionInvokerCache controllerActionInvokerCache,
+            IOptions<MvcOptions> optionsAccessor,
+            ILoggerFactory loggerFactory,
+            DiagnosticListener diagnosticListener,
+            IActionResultTypeMapper mapper,
+            IActionContextAccessor actionContextAccessor)
+        {
+            _controllerActionInvokerCache = controllerActionInvokerCache;
+            _valueProviderFactories = optionsAccessor.Value.ValueProviderFactories.ToArray();
+            _maxModelValidationErrors = optionsAccessor.Value.MaxModelValidationErrors;
+            _enableActionInvokers = optionsAccessor.Value.EnableActionInvokers;
+            _logger = loggerFactory.CreateLogger<ControllerActionInvoker>();
+            _diagnosticListener = diagnosticListener;
+            _mapper = mapper;
+            _actionContextAccessor = actionContextAccessor ?? ActionContextAccessor.Null;
+        }
+
+        public RequestDelegate CreateRequestDelegate(ActionDescriptor actionDescriptor, RouteValueDictionary dataTokens)
+        {
+            // Fallback to action invoker extensibility so that invokers can override any default behaviors
+            if (!_enableActionInvokers && actionDescriptor is ControllerActionDescriptor)
+            {
+                return context =>
+                {
+                    RouteData routeData = null;
+
+                    if (dataTokens is null or { Count: 0 })
+                    {
+                        routeData = new RouteData(context.Request.RouteValues);
+                    }
+                    else
+                    {
+                        routeData = new RouteData();
+                        routeData.PushState(router: null, context.Request.RouteValues, dataTokens);
+                    }
+
+                    var actionContext = new ActionContext(context, routeData, actionDescriptor);
+
+                    var controllerContext = new ControllerContext(actionContext)
+                    {
+                        // PERF: These are rarely going to be changed, so let's go copy-on-write.
+                        ValueProviderFactories = new CopyOnWriteList<IValueProviderFactory>(_valueProviderFactories)
+                    };
+
+                    controllerContext.ModelState.MaxAllowedErrors = _maxModelValidationErrors;
+
+                    var (cacheEntry, filters) = _controllerActionInvokerCache.GetCachedResult(controllerContext);
```

Comment:
Another thing to investigate. This is a cache lookup that hits a concurrent dictionary (keyed on ActionDescriptor). Since you're creating a request delegate per-action there's no need for additional caches and all that stuff can get closed over here.

Reply:
https://github.com/dotnet/aspnetcore/pull/27763